File: app/services/plans.py
"""What each plan admits, in one place.

The plan's instruction for this phase is the whole design: do not
hard-code plan checks around the codebase, create centralised capability
checks. So a plan is not a row that somebody can edit into a different
product -- it is code, here, and the subscription table only records which
of these a workspace is on.

Two kinds of thing, and they are enforced differently on purpose. A
*feature* is a yes or no, checked once at the door by a dependency in a
route's signature. A *limit* is a number, checked against what a workspace
already has, which means counting something -- and counting is a query, so
those live in the service.

Prices here are what a plan costs, for the page that lists them. What the
payment provider calls that price is configuration, because it is the
provider's identifier and not this product's fact.
"""

from dataclasses import dataclass
from decimal import Decimal
from enum import StrEnum

# Null in a limit means no ceiling. Spelled as a constant because
# `limits[X] is None` reads as "missing" at a glance and means the
# opposite.
UNLIMITED = None


class PlanTier(StrEnum):
    """The three the plan proposes, in the order they grow."""

    STARTER = "starter"
    GROWTH = "growth"
    BUSINESS = "business"


class Feature(StrEnum):
    """Something a plan either admits or does not.

    Checked at the door rather than counted. Every one of these is a
    capability a route can be gated on with one line in its signature,
    which is what stops a plan check from being an `if` somebody forgets
    to write in the route they add next month.
    """

    AUTOMATIONS = "automations"
    ECOMMERCE = "ecommerce"
    ADVANCED_ANALYTICS = "advanced_analytics"
    API_ACCESS = "api_access"
    AUDIT_LOGS = "audit_logs"


class PlanLimit(StrEnum):
    """Something a plan allows a number of.

    Enforced by counting what a workspace already has, which is why these
    are checked in the service and not in a signature: a route cannot be
    told "at most ten" without somebody running a query.
    """

    WHATSAPP_NUMBERS = "whatsapp_numbers"
    TEAM_MEMBERS = "team_members"
    AI_RESPONSES_PER_MONTH = "ai_responses_per_month"
    KNOWLEDGE_DOCUMENTS = "knowledge_documents"


@dataclass(frozen=True)
class Plan:
    tier: PlanTier
    name: str
    description: str
    price: Decimal
    currency: str
    features: frozenset[Feature]
    limits: dict[PlanLimit, int | None]

    def admits(self, feature: Feature) -> bool:
        return feature in self.features

    def ceiling(self, limit: PlanLimit) -> int | None:
        return self.limits.get(limit, UNLIMITED)

    @property
    def is_free(self) -> bool:
        return self.price == 0


PLANS: dict[PlanTier, Plan] = {
    PlanTier.STARTER: Plan(
        tier=PlanTier.STARTER,
        name="Starter",
        description="One number, a couple of people, and the assistant.",
        # Free, and that matters structurally rather than commercially:
        # it is what a workspace has before it has ever paid, and what it
        # falls back to when a payment stops working. A product whose
        # free tier does nothing would lock a business out of its own
        # inbox over a declined card.
        price=Decimal("0"),
        currency="USD",
        features=frozenset(),
        limits={
            PlanLimit.WHATSAPP_NUMBERS: 1,
            PlanLimit.TEAM_MEMBERS: 2,
            PlanLimit.AI_RESPONSES_PER_MONTH: 1_000,
            PlanLimit.KNOWLEDGE_DOCUMENTS: 50,
        },
    ),
    PlanTier.GROWTH: Plan(
        tier=PlanTier.GROWTH,
        name="Growth",
        description="A team, a storefront, and automations.",
        price=Decimal("49"),
        currency="USD",
        features=frozenset(
            {
                Feature.AUTOMATIONS,
                Feature.ECOMMERCE,
                Feature.ADVANCED_ANALYTICS,
            }
        ),
        limits={
            PlanLimit.WHATSAPP_NUMBERS: 3,
            PlanLimit.TEAM_MEMBERS: 10,
            PlanLimit.AI_RESPONSES_PER_MONTH: 10_000,
            PlanLimit.KNOWLEDGE_DOCUMENTS: 500,
        },
    ),
    PlanTier.BUSINESS: Plan(
        tier=PlanTier.BUSINESS,
        name="Business",
        description="High volume, several stores, and the API.",
        price=Decimal("199"),
        currency="USD",
        features=frozenset(Feature),
        limits={
            PlanLimit.WHATSAPP_NUMBERS: UNLIMITED,
            PlanLimit.TEAM_MEMBERS: UNLIMITED,
            PlanLimit.AI_RESPONSES_PER_MONTH: 100_000,
            PlanLimit.KNOWLEDGE_DOCUMENTS: UNLIMITED,
        },
    ),
}

# What a workspace is on before it has ever subscribed, and what it falls
# back to when a subscription stops being good. Named rather than written
# as PLANS[PlanTier.STARTER] in three places, so that changing which plan
# is the floor is one edit.
FREE_PLAN = PLANS[PlanTier.STARTER]

File: web/lib/plans.ts
/**
 * What a workspace may do, and how to say it.
 *
 * Pure, so a client component can import it. The rule below is the one
 * every other screen in this client depends on, and it is short enough to
 * state in a sentence: **gate on `plan`, never on `subscription.plan`.**
 *
 * They disagree routinely and the disagreement is the point. A `past_due`
 * subscription still entitles the full plan while the provider retries; a
 * workspace the platform has comped is entitled to more than it pays for.
 * A client that gated on what is being paid for would take features away
 * from a customer whose card is merely being retried, which is the exact
 * failure the API arranged its response shape to prevent.
 */

import type {
  Feature,
  MetricUsage,
  Plan,
  PlanLimit,
  PlanTier,
  SubscriptionStatus,
  UsageMetric,
  WorkspacePlan,
} from "@/lib/types";

/** Whether this workspace may use a capability right now. */
export function admits(current: WorkspacePlan | null, feature: Feature): boolean {
  return current?.plan.features.includes(feature) ?? false;
}

/** What the plan allows of something. `null` is unlimited, never zero. */
export function ceiling(
  current: WorkspacePlan | null,
  limit: PlanLimit,
): number | null {
  return current?.plan.limits[limit] ?? null;
}

/**
 * The statuses under which a workspace actually gets what it pays for.
 *
 * `past_due` is in here deliberately, matching the API: a card that did
 * not go through is a provider still retrying, and taking a business's
 * features away over a bank's fraud check is the wrong way to lose a
 * customer. What happens instead is that somebody is told.
 */
const ENTITLING: SubscriptionStatus[] = ["active", "trialing", "past_due"];

export function isEntitling(status: SubscriptionStatus): boolean {
  return ENTITLING.includes(status);
}

/**
 * Whether to warn about the subscription, without disabling anything.
 *
 * `past_due` and `unpaid` both want a warning; only the first still
 * entitles. The screens read these separately for that reason.
 */
export function needsAttention(current: WorkspacePlan | null): boolean {
  const status = current?.subscription?.status;

  return status === "past_due" || status === "unpaid";
}

/**
 * Stopping, as distinct from stopped.
 *
 * Cancelling sets `cancel_at_period_end` and leaves the status `active`,
 * because the month has been paid for. A screen that could not tell these
 * apart would say the wrong thing about both.
 */
export function isEnding(current: WorkspacePlan | null): boolean {
  const subscription = current?.subscription;

  return subscription?.cancel_at_period_end === true && subscription.status === "active";
}

export const TIER_ORDER: PlanTier[] = ["starter", "growth", "business"];

export function isUpgradeFrom(from: PlanTier, to: PlanTier): boolean {
  return TIER_ORDER.indexOf(to) > TIER_ORDER.indexOf(from);
}

/**
 * A plan with nothing to pay. Cancelling is how somebody returns to it.
 *
 * Tested against the string rather than parsed, like everything else that
 * touches money here: `"0"`, `"0.00"` and `"0.000"` are all the free plan.
 */
export function isFree(plan: Plan): boolean {
  return /^0(\.0+)?$/.test(plan.price);
}

// --- wording ------------------------------------------------------------

export const FEATURE_LABEL: Record<Feature, string> = {
  automations: "Automations",
  ecommerce: "Storefront sync",
  advanced_analytics: "Advanced analytics",
  api_access: "API access",
  audit_logs: "Audit log",
};

export const LIMIT_LABEL: Record<PlanLimit, string> = {
  whatsapp_numbers: "WhatsApp numbers",
  team_members: "Team members",
  ai_responses_per_month: "Assistant replies a month",
  knowledge_documents: "Knowledge documents",
};

export const METRIC_LABEL: Record<UsageMetric, string> = {
  ai_responses: "Assistant replies",
  ai_tokens: "Assistant tokens",
  whatsapp_messages: "WhatsApp messages",
  active_contacts: "Contacts",
  team_members: "Team members",
  whatsapp_numbers: "WhatsApp numbers",
  knowledge_documents: "Knowledge documents",
  knowledge_tokens: "Knowledge tokens",
};

/** Unlimited, spelled out. A blank cell reads as a missing value. */
export function ceilingLabel(limit: number | null): string {
  return limit === null ? "Unlimited" : limit.toLocaleString();
}

/**
 * How full a meter is, as a percentage, or null where nothing bounds it.
 *
 * Integers, so no float creeps into a number somebody reads. A limit of
 * zero would divide by nothing, so it reads as full rather than as an
 * error.
 */
export function fractionUsed(usage: MetricUsage): number | null {
  if (usage.limit === null) return null;
  if (usage.limit === 0) return 100;

  return Math.min(100, Math.round((usage.quantity / usage.limit) * 100));
}

/** The codes that mean the plan is what is in the way, rather than you. */
const PLAN_CODES = ["feature_not_in_plan", "plan_limit_reached"];

export function isPlanRefusal(code: string | undefined): boolean {
  return code !== undefined && PLAN_CODES.includes(code);
}
